jme3-core/src/main/java/com/jme3/app/state/ScreenshotAppState.java

```diff
@@ -187,6 +195,26 @@ public void initialize(AppStateManager stateManager, Application app) {
         super.initialize(stateManager, app);
     }
 
+    /**
+     * Clean up this AppState during the first update after it gets detached.
+     */
+    @Override
+    public void cleanup() {
+        if (inputManager != null) {
+            inputManager.deleteMapping("ScreenShot");
+            inputManager.removeListener(this);
+            inputManager = null;
+        }
+
+        ViewPort viewPort = last;

```

Comment:
Just so I get this, you use `viewPort` because calling `viewPort.removeProcessor` triggers a cleanup call on this AppState?
Because otherwise this wouldn't be needed, but I am not 100% certain what you mean with indirect recursion, might need some more commenting.

Also is there a reason why super.cleanup() is called last and not first? I guess it doesn't make a difference, but I am used to it being the first statement.

Reply:
Yes, `removeProcessor()` calls `cleanup()`. So if we weren't careful, we'd get an infinite recursion. I should document that fact explicitly, I guess.

My view is that subclasses are built upon (and depend on) superclasses. When I'm initializing, I initialize the superclass before the subclass. Then, when I'm cleaning up, I clean up the subclass before the superclass.

BTW, due to the recursion, it's likely that `super.cleanup()` gets invoked twice, but that doesn't seem to harm anything.

Reply:
I've added some javadoc explaining the recursion hazard in detail.